File: 0211-design-add-and-search-words-data-structure/0211-design-add-and-search-words-data-structure.ts
class TrieNode {
    children: { [key: string]: TrieNode };
    isEndOfWord: boolean;

    constructor() {
        this.children = {};
        this.isEndOfWord = false;
    }
}

class Trie {
    root: TrieNode;

    constructor() {
        this.root = new TrieNode();
    }

    insert(word: string): void {
        let current = this.root;
        for (const c of word) {
            if (!(c in current.children)) {
                current.children[c] = new TrieNode();
            }
            current = current.children[c];
        }
        current.isEndOfWord = true;
    }

    search(word: string): boolean {
        return this._search(word, this.root);
    }

    private _search(word: string, node: TrieNode): boolean {
        for (let i = 0; i < word.length; i++) {
            const c = word[i];
            if (c === ".") {
                for (const child of Object.values(node.children)) {
                    if (this._search(word.slice(i + 1), child)) {
                        return true;
                    }
                }
                return false;
            } else {
                if (!(c in node.children)) {
                    return false;
                }
                node = node.children[c];
            }
        }
        return node.isEndOfWord;
    }

    startsWith(prefix: string): boolean {
        let current = this.root;
        for (const c of prefix) {
            if (!(c in current.children)) {
                return false;
            }
            current = current.children[c];
        }
        return true;
    }
}

class WordDictionary {
    private trie: Trie;

    constructor() {
        this.trie = new Trie();
    }

    addWord(word: string): void {
        this.trie.insert(word);
    }

    search(word: string): boolean {
        return this.trie.search(word);
    }
}

File: 0211-design-add-and-search-words-data-structure/0211-design-add-and-search-words-data-structure.py
class TrieNode:
    def __init__(self):
        self.children = {}  # dictionary: char -> TrieNode
        self.is_end_of_word = False


class Trie:
    def __init__(self):
        self.root = TrieNode()

    def insert(self, word: str) -> None:
        current = self.root
        for c in word:
            if c not in current.children:
                current.children[c] = TrieNode()
            current = current.children[c]
        current.is_end_of_word = True

    def search(self, word: str) -> bool:
        return self._search(word, self.root)

    def _search(self, word: str, node: TrieNode) -> bool:
        for i, c in enumerate(word):
            if c == '.':
                for child in node.children.values():
                    if self._search(word[i + 1:], child):
                        return True
                return False
            else:
                if c not in node.children:
                    return False
                node = node.children[c]
        return node.is_end_of_word

    def starts_with(self, prefix: str) -> bool:
        current = self.root
        for c in prefix:
            if c not in current.children:
                return False
            current = current.children[c]
        return True


class WordDictionary:
    def __init__(self):
        self.trie = Trie()

    def addWord(self, word: str) -> None:
        self.trie.insert(word)

    def search(self, word: str) -> bool:
        return self.trie.search(word)
